README.md
# Format Preserving Encryption in Node

An implementation of the NIST-approved FF1 and FF3-1 algorithms in Node.

This implementation conforms (as best as possible) to
[Draft SP 800-38G Rev. 1][800-38g1]. The implementation passes all tests
specified by NIST in their Cryptographic Standards and Guidelines
[examples for FF1][ff1-examples]; however, no official examples/samples exist
(or are known) for FF3-1. FF3 is not implemented as NIST has officially
deprecated its use in light of recent [cryptanalysis][ff3-cryptanalysis]
performed on it.


## Installation

#### Using the npm or yarn package managers:
You may want to make sure you are running the latest version of npm or yarn by first executing
```sh
$ npm install -g npm
# or
$ npm install -g yarn
```

Install the ubiq-security-fpe package with:

```sh
$ npm install ubiq-security-fpe
# or
$ yarn add ubiq-security-fpe
```

To build and install directly from a clone of the gitlab repository source:

```sh
$ git clone git@gitlab.com:ubiqsecurity/ubiq-fpe-node.git
$ cd ubiq-node-fpe
$ npm install
```

### Requirements

Node.js version 12 or later
npm version 6 or later

All dependencies are pre-required in the module itself.




# Testing

To run the tests:
```sh
$ npm run test
```

As described above, the unit tests for FF1 come from the NIST guidelines. As
no such guidelines are available for FF3-1, the unit tests verify only that
the encryption and decryption implementations are compatible with each other.

# Documentation

See the [Node API docs][apidocs].




### About alphabets and the radix parameter

The interfaces operate on strings, and the radix parameter determines which
characters are valid within those strings, i.e. the alphabet. For example, if
your radix is 10, then the alphabet for your plain text consists of the
characters in the string "0123456789". If your radix is 16, then the
alphabet is the characters in the string "0123456789abcdef".

A radix of up to 255 is supported, and the alphabet for a radix of 36 is
"0123456789abcdefghijklmnopqrstuvwxyz".

### Tweaks

Tweaks are very much like Initialization Vectors (IVs) in "traditional"
encryption algorithms. For FF1, the minimun and maximum allowed lengths of
the tweak may be specified by the user, and any tweak length between those
values may be used. For FF3-1, the size of the tweak is fixed at 7 bytes.

### Plain/ciphertext input lengths

For both FF1 and FF3-1, the minimum length is determined by the inequality:
- radix<sup>minlen</sup> >= 1000000

or:
- minlen >= 6 / log<sub>10</sub> radix

Thus, the minimum length is determined by the radix and is automatically
calculated from it.

For FF1, the maximum input length is
- 2<sup>32</sup>

For FF3-1, the maximum input length is
- 2 * log<sub>radix</sub> 2<sup>96</sup>

or:
- 192 / log<sub>2</sub> radix

## Examples

The [unit test code](test) provides the best and simplest example of how to use the
interfaces.


### FF1
```javascript
   /*
     * @key is an array of bytes which must be 16, 24, or 32 long
     * @twk is an array of bytes of which must be between @twkmin
     *     and @twkmax lengths passed in. 
     * @custom_radix_str is the custom radix string indicating valid
     *     characters allowed by @PT and @encrypted_text.  In this
     *     case, Radix 10 (numeric) is being used.
     * 
     * @PT is "user input" for the data being encrypted
     */

    const custom_radix_str = "0123456789"

    const ctx = new FF1(key, twk, twkmin, twkmax, custom_radix_str.length, custom_radix_str);
    const cipher_text = ctx.encrypt(PT);
    const decrypted_text = ctx.decrypt(cipher_text);
```

[800-38g1]:https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-38Gr1-draft.pdf
[ff1-examples]:https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/FF1samples.pdf
[ff3-cryptanalysis]:https://csrc.nist.gov/News/2017/Recent-Cryptanalysis-of-FF3
[dashboard]:https://dashboard.ubiqsecurity.com
[credentials]:https://dev.ubiqsecurity.com/docs/how-to-create-api-keys
[apidocs]:https://dev.ubiqsecurity.com/docs/api



CHANGELOG.md
## 2.0.0 - 2023-04-20
* Update version
* Update referenced packages based on vulnerabilities

## 1.0.1 - 2022-06-24
* Update to README

## 1.0.1 - 2022-06-21
* Initial Version



test/Bn.test.js
const bigInt = require('big-integer');
const bn = require('../lib/Bn');

/*
12/20/21
Check in initial set of unit tests.

How to execute:

$ npm run test

> ubiq-fpe-node@1.0.0 test /Users/anthonyiasi/ubiq-fpe-node
> jest

 PASS  test/Bn.test.js
  ✓ radix_exceptions (10 ms)
  ✓ radix_edgecase (1 ms)
  ✓ radix_dec2hex (1 ms)
  ✓ radix_oct2hex
  ✓ radix_dec2dec (1 ms)
  ✓ radix_oct2dec

Test Suites: 1 passed, 1 total
Tests:       6 passed, 6 total
Snapshots:   0 total
Time:        0.258 s, estimated 1 s
Ran all test suites.
*/

test('radix_exceptions', () => {
    expect(() => bn.bigint_set_str('109', '012345678')).toThrow();
    expect(() => bn.bigint_set_str('109', '')).toThrow();
    expect(() => bn.bigint_get_str('', 0)).toThrow();
});

test('radix_edgecase', () => {
    const r1 = bn.bigint_set_str('0', '0123456789');
    expect(r1.value).toBe(bigInt(0).value);
    // expect(bn.bigint_get_str('0123456789ABCDEF', r1.value).value).toBe(bigInt(0).value); // todo enable
    expect(bn.bigint_get_str('0123456789ABCDEF', 0)).toBe('0');
});

test('radix_dec2hex', () => {
    const r1 = bn.bigint_set_str('100', '0123456789');
    expect(r1.value).toBe(bigInt(100).value);
    expect(bn.bigint_get_str('0123456789ABCDEF', r1)).toBe('64');
});

test('radix_oct2hex', () => {
    const r1 = bn.bigint_set_str('100', '01234567');
    expect(r1.value).toBe(bigInt(64).value);
    expect(bn.bigint_get_str('0123456789ABCDEF', r1)).toBe('40');
});

test('radix_dec2dec', () => {
    const r1 = bn.bigint_set_str('@$#', '!@#$%^&*()');
    expect(r1.value).toEqual(bigInt(132).value);
    expect(bn.bigint_get_str('0123456789', r1)).toBe('132');
});

test('radix_oct2dec', () => {
    const r1 = bn.bigint_set_str('@$#', '!@#$%^&*');
    expect(r1.value).toEqual(bigInt(90).value);
    expect(bn.bigint_get_str('0123456789', r1)).toEqual('90');
});

test('bigint so_alphanum_po', () => {
    const r1 = bn.bigint_set_str('1234', ' 0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ');
    expect(r1.value).toEqual(bigInt(105566).value);
    expect(bn.bigint_get_str(' 0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ', '105566')).toEqual('1234');
});


test/FF3_1.testx.js
const assert = require('assert');
const FF3_1 = require('../lib/FF3_1');

function test(key, twk, PT, CT, radix) {
    let out = String();
    ctx = FF3_1;

    assert.equal(PT.length, CT.length);

    ctx = new FF3_1(key, twk, radix);

    out = ctx.encrypt(PT);
    assert.equal(CT, out);

    out = ctx.decrypt(CT);
    assert.equal(PT, out);
}

const key = new Uint8Array(
    [0xef, 0x43, 0x59, 0xd8,
        0xd5, 0x80, 0xaa, 0x4f,
        0x7f, 0x03, 0x6d, 0x6f,
        0x04, 0xfc, 0x6a, 0x94,
        0x3b, 0x80, 0x6a, 0xeb,
        0x63, 0x08, 0x27, 0x1f,
        0x65, 0xcf, 0x33, 0xc7,
        0x39, 0x1b, 0x27, 0xf7],
);

const twk1 = new Uint8Array(
    [0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00],
);

const twk2 = new Uint8Array(
    [0x39, 0x38, 0x37, 0x36,
        0x35, 0x34, 0x33],
);

const twk3 = new Uint8Array(
    [0x37, 0x37, 0x37, 0x37,
        0x70, 0x71, 0x72],
);

const _pt = ['890121234567890000', '89012123456789abcde'];

test(
    key.slice(0, 16),
    twk1,
    _pt[0],

    '075870132022772250',

    10,
);


Bn.js
var bigInt = require("big-integer");



/**
 * Algorithms to convert a numerical value in a given alphabet to a number
 */
class Bn
{
	/**
     * Convert a numerical value in a given alphabet to a number.
     *
     * An alphabet consists of single-byte symbols in which each
     * symbol represents the numerical value associated with its
     * index/position in the alphabet. for example, consider the
     * alphabet: !@#$%^-*()
     * In this alphabet ! occupies index 0 and is therefore
     * assigned that value. @ = 1, # = 2, etc. Furthermore, the
     * alphabet contains 10 characters, so that becomes the radix
     * of the input. Using the alphabet above, an input of @$#
     * translates to a value of 132 (one hundred thirty-two,
     * decimal).
     *
     * If the alphabet above were instead: !@#$%^-*
     * The radix would be 8 and an input of @$# translates to a
     * value of 90 (ninety, decimal).
     *
     * @param str the numerical value to be converted
     * @param alpha alphabet consists of single-byte symbols
     *
     * @return the numerical value of the str pattern 
     * position found in the alphabet
     */
	
    static __bigint_set_str(str, alpha) {
        var len = str.length;
		/*
		 * the alphabet can be anything and doesn't have
		 * to be in a recognized canonical order. the only
		 * requirement is that every value in the list be
		 * unique. checking that constraint is an expensive
		 * undertaking, so it is assumed. as such, the radix
		 * is simply the number of characters in the alphabet.
		 */
		var rad = alpha.length;
		if (rad <= 0)
		{
			throw new Error("invalid argument, alphabet cannot be empty");
		}

        var a = bigInt();
		var i = 0;
		/* represents the numerical value of str */
        var x = bigInt(0);
		/*
		 * multiplier used to multiply each digit
		 * of the input into its correct position
		 */
        var m = bigInt(1);

		for (i = 0; i < len; i++)
		{
			var pos = 0;
			/*
			 * determine index/position in the alphabet.
			 * if the character is not present the input
			 * is not valid.
			 */
			pos = alpha.indexOf(str.charAt(len - 1 - i));
			if (pos < 0)
			{
				throw new Error("invalid argument, input character not found in alphabet");
			}
			/*
			 * multiply the digit into the correct position
			 * and add it to the result
			 */ 
            a = m * bigInt(pos);
			x = x + a;
			m = m * bigInt(rad);
		}
		return x;
    }





    /**
     * Inserts a character at a position in a String.
     *
     * Convenience function returns String with inserted char 
     * at an index position.
     *
     * @param str the original String
     * @param ch the character to insert
     * @param position the index position where to insert the ch
     *
     * @return    the new String containing the inserted ch 
     */ 
    static insertChar(str, ch, position) {
		return str.substring(0, position) + ch + str.substr(position);
	}


 
     


    /**
     * Gets the str pattern of the alphabet given the numeric value.
     *
     * @param alpha alphabet consists of single-byte symbols
     * @param x the numerical value of the str pattern
     *
     * @return the new String of the converted value 
     */ 
    static __bigint_get_str(alpha, x) {
		var rad = alpha.length;
        var quotient = bigInt();
		quotient = x;
		var str = "";

		if (rad <= 0)
		{
			throw new Error("invalid argument, alphabet cannot be empty");
		}

		/*
		 * to convert the numerical value, repeatedly
		 * divide (storing the resulted quotient and the remainder)
		 * by the desired radix of the output.
		 *
		 * the remainder is the current digit; the result
		 * of the division becomes the input to the next
		 * iteration
		 */
        while (bigInt(quotient).compareTo(bigInt(0)) != 0)
		{
			var remainder = 0;
			var result = bigInt(quotient).divmod(bigInt(rad));
            remainder = result.remainder;
            quotient = result.quotient;
			str = Bn.insertChar(str, alpha.charAt(remainder), 0);
		}

		if (str.length == 0)
		{
			str = Bn.insertChar(str, alpha.charAt(0), 0);
		}
		return str;
	}

}


/* make available to other modules */
exports.__bigint_set_str = Bn.__bigint_set_str;
exports.__bigint_get_str = Bn.__bigint_get_str;






lib/FF1.js
/* eslint-disable no-bitwise */

const bigInt = require('big-integer');
const FFX = require('./FFX');
const arrayUtil = require('./arrayUtil');
const Bn = require('./Bn');

function bytesToInteger(hex, d) {
	let pow = bigInt.one;
	let result = bigInt.zero;
	for (let i = 1; i <= d; i++) {
		result = result.add(pow.multiply(hex[d - i]));
		pow = pow.multiply(256);
	}
	return result;
}
/*
12/27/21
Check in non-functional initial code skeleton.

How to compile:
node FF1.js
*/

class FF1 extends FFX {
	constructor(key, twk, twkmin, twkmax, radix, custom_radix_str) {
		if (custom_radix_str) {
			radix = custom_radix_str.length;
		}
		super(key, twk, bigInt.one.shiftLeft(32), twkmin, twkmax, radix, custom_radix_str);
	}

	/*
	 * The comments below reference the steps of the algorithm described here:
	 *
	 * https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-38Gr1-draft.pdf
	 */
	cipher(X, twk, encrypt) {
		/* Step 1 */
		const n = X.length;
		// Let u = ⎣n/2⎦; v = n – u.
		const u = Math.floor(parseInt(n / 2, 10));
		const v = n - u;

		/* Step 2 */
		let A = null;
		let B = null;
		// Let A = X[1..u]; B = X[u + 1..n].
		if (encrypt) {
			A = X.substring(0, u);
			B = X.substring(u);
		} else {
			B = X.substring(0, u);
			A = X.substring(u);
		}
		/* Step 3 */
		// 3.	 Let b = ⎡ ⎡v⋅LOG(radix)⎤/8⎤.
		const x = (parseInt(Math.ceil((Math.log(this.radix) / Math.log(2)) * v), 10) + 7);
		const b = parseInt(x / 8, 10);

		/* Step 4 */
		// 4.	 Let d = 4⎡b/4⎤ + 4.
		const d = 4 * (parseInt((b + 3) / 4, 10)) + 4;
		const p = 16;
		const r = (parseInt((d + 15) / 16, 10)) * 16;
		const R = new Uint8Array(r).fill(null);
		let q = 0;
		/* use default tweak if none is supplied */
		if (!twk) {
			twk = this.twk;
		}
		/* check text and tweak lengths */
		if (n < this.txtmin || n > this.txtmax) {
			throw new Error('invalid input length');
		} else if (twk.length < this.twkmin || (this.twkmax > 0 && twk.length > this.twkmax)) {
			throw new Error('invalid tweak length');
		}
		/* the number of bytes in Q */
		q = (parseInt((twk.length + b + 1 + 15) / 16, 10)) * 16;
		/*
		 * P and Q need to be adjacent in memory for the
		 * purposes of encryption
		 */
		const PQ = new Uint8Array(p + q).fill(null);
		PQ[0] = 1;
		PQ[1] = 2;
		PQ[2] = 1;
		PQ[3] = this.radix >> 16;
		PQ[4] = this.radix >> 8;
		PQ[5] = this.radix >> 0;
		PQ[6] = 10;
		PQ[7] = u;
		PQ[8] = n >> 24;
		PQ[9] = n >> 16;
		PQ[10] = n >> 8;
		PQ[11] = n >> 0;
		PQ[12] = twk.length >> 24;
		PQ[13] = twk.length >> 16;
		PQ[14] = twk.length >> 8;
		PQ[15] = twk.length >> 0;
		/* Step 6i, the static parts */
		arrayUtil.arrayCopy(twk, 0, PQ, p, twk.length);
		/* remainder of Q already initialized to 0 */
		for (let i = 0; i < 10; i++) {
			/* Step 6v */
			const m = ((i + !!encrypt) % 2) ? u : v;

			const radixPow = bigInt(this.radix).pow(m);
			let c = null;
			let y = null;
			let numb;
			/* Step 6i, the non-static parts */
			PQ[PQ.length - b - 1] = encrypt ? i : (9 - i);
			/*
			* convert the numeral string B to an integer and
			* export that integer as a byte array into Q
			*/
			c = Bn.bigint_set_str(B, this.custom_radix_str, this.radix);// bigInt(B, this.radix);
			numb = this.BigIntToByteArray(c);
			if (numb[0] === 0 && numb.length > 1) {
				/*
				 * Per the Java documentation, BigInteger.toByteArray always
				 * returns enough bytes to contain a sign bit. For the purposes
				 * of this function all numbers are unsigned; however, when the
				 * most-significant bit is set in a number, the Java library
				 * returns an extra most-significant byte that is set to 0.
				 * That byte must be removed for the cipher to work correctly.
				 */
				numb = arrayUtil.copyOfRange(numb, 1, numb.length);
			}
			if (b <= numb.length) {
				arrayUtil.arrayCopy(numb, 0, PQ, PQ.length - b, b);
			} else {
				/* pad on the left with zeros */
				PQ.fill(0, PQ.length - b, PQ.length - numb.length);
				arrayUtil.arrayCopy(numb, 0, PQ, PQ.length - numb.length, numb.length);
			}
			/* Step 6ii */
			this.prf(R, 0, PQ, 0, PQ.length);
			/*
			 * Step 6iii
			 * if r is greater than 16, fill the subsequent blocks
			 * with the result of ciph(R ^ 1), ciph(R ^ 2), ...
			 */
			// Let S be the first d bytes of the following string of ⎡d/16⎤ blocks:
			// R || CIPHK (R ⊕ [1]16) || CIPHK (R ⊕ [2]16) … CIPHK (R ⊕ [⎡d/16⎤–1]16).
			let l;
			for (let j = 1; j < parseInt(r / 16, 10); j++) {
				l = j * 16;
				R.fill(0, l, l + 12);
				R[l + 12] = j >> 24;
				R[l + 13] = j >> 16;
				R[l + 14] = j >> 8;
				R[l + 15] = j >> 0;

				this.xor(R, l, R, 0, R, l, 16);
				this.ciph(R, l, R, l);
			}
			/*
			 * Step 6vi
			 * calculate A +/- y mod radix**m
			 * where y is the number formed by the first d bytes of R
			 * * create an integer from the first @d bytes in @R
			 */
			// vi.	 Let c = (NUM radix (A)+y) mod radix m .
			c = Bn.bigint_set_str(A, this.custom_radix_str, this.radix);
			y = bytesToInteger(R, d);
			y = y.mod(radixPow);
			// vii.	 Let C = STR m	radix(c).
			if (encrypt) {
				c = c.add(y);
			} else {
				c = c.subtract(y);
			}

			c = c.mod(radixPow);
			// the algorithm appears to need a number between 0 and the dominator,
			// this if statement prevents result to be negative.
			if (c < 0) {
				c = c.add(radixPow);
			}
			const C = this.str(m, this.radix, c);
			/* Step 6viii */
			A = B;
			B = C;
		}
		/* Step 7 */
		return encrypt ? (A + B) : (B + A);
	}
}
// /* make available to other modules */
module.exports = FF1;


lib/FFX.js
/* eslint-disable no-bitwise */

const bigInt = require('big-integer');
const { createCipheriv } = require('crypto');
const errorMessages = require('./errorMessages');
const { arrayCopy } = require('./arrayUtil');
const Bn = require('./Bn');
/*
12/27/21
Check in non-functional initial code skeleton.
How to compile:
node FFX.js
*/

class FFX {
    constructor(key, twk, txtmax, twkmin, twkmax, radix, custom_radix_str) {
        let txtmin = 0;
        this.radix = 0;
        this.txtmin = 0;
        this.txtmax = 0;
        this.twkmin = 0;
        this.twkmax = 0;
        this.twk = [];
        this.custom_radix_str = undefined;

        // this.cipher = null;
        let algorithm;
        /* all 3 key sizes of AES are supported */
        switch (key.length) {
            case 16:
                algorithm = 'aes-128-cbc';
                break;
            case 24:
                algorithm = 'aes-192-cbc';
                break;
            case 32:
                algorithm = 'aes-256-cbc';
                break;
            default:
                throw new Error(`key size error: ${key.length}`);
        }
        /*
         * FF1 and FF3-1 support a radix up to 65536, but the
         * implementation becomes increasingly difficult and
         * less useful in practice after the limits below.
         */
        if (radix < 2 || radix > 36) {
            // throw new Error('invalid radix');
        }

        /*
         * for both ff1 and ff3-1: radix**minlen >= 1000000
         *
         * therefore:
         *   minlen = ceil(log_radix(1000000))
         *          = ceil(log_10(1000000) / log_10(radix))
         *          = ceil(6 / log_10(radix))
         */
        txtmin = parseInt(Math.ceil(6.0 / Math.log10(radix), 10), 10);
        if (txtmin < 2 || txtmin > txtmax) {
            throw new Error('minimum text length out of range');
        }

        /* the default tweak must be specified */
        if (!twk) {
            throw new Error('invalid tweak');
        }
        /* check tweak lengths */
        if (twkmin > twkmax || twk.length < twkmin || (twkmax > 0 && twk.length > twkmax)) {
            throw new Error('invalid tweak length');
        }

        const iv = Buffer.alloc(16, 0);
        const keyArrayBuffer = new Buffer.from(key);
        this.algorithm = algorithm;
        this.keyArrayBuffer = keyArrayBuffer;
        this.iv = iv;
        this.Cipher = createCipheriv(algorithm, keyArrayBuffer, iv);
        this.Cipher.setAutoPadding(false);

        this.radix = radix;
        this.txtmin = txtmin;
        this.txtmax = txtmax;
        this.twkmin = twkmin;
        this.twkmax = twkmax;
        this.twk = [...twk];
        this.custom_radix_str = custom_radix_str;
    }

    /*
     * perform an aes-cbc encryption (with an IV of 0) of @src, storing
     * the last block of output into @dst. The number of bytes in @src
     * must be a multiple of 16. @dst and @src may point to the same
     * location but may not overlap, otherwise. @dst must point to a
     * location at least 16 bytes long
     */
    prf(dst, doff, src, soff, len) {
        const blksz = 16; // Should be something like this.Cipher.getBlockSize or getCiherInfo.block size
        if ((src.length - soff) % blksz !== 0) {
            throw new Error('invalid source length');
        }
        let tempResult;
        for (let i = 0; i < len && i < src.length - soff; i += blksz) {
            const temp = new Uint8Array(blksz);
            arrayCopy(src, soff + i, temp, 0, blksz);
            tempResult = this.Cipher.update(temp);
        }
        this.Cipher.final();

        for (let j = 0; j < tempResult.length; j++) {
            dst[j + doff] = tempResult[j];
        }
        // reset Cipher
        this.Cipher = createCipheriv(this.algorithm, this.keyArrayBuffer, this.iv);
        this.Cipher.setAutoPadding(false);
    }

    /*
     * perform an aes-ecb encryption of @src. @src and @dst must each be
     * 16 bytes long, starting from the respective offsets. @src and @dst
     * may point to the same location or otherwise overlap
     */
    ciphh(dst, doff, src, soff) {
        this.prf(dst, doff, src, soff, 16);
    }

    /*
     * a convenience version of the ciph function that returns its
     * output as a separate byte array
     */
    ciph(dst, doff, src, off) {
        this.prf(dst, doff, src, off, 16);
    }

    /*
     * convenience function that returns the reversed sequence
     * of bytes as a new byte array
     */
    rev(src) {
        const dst = [...src];
        return dst.reverse();
    }

    /*
     * reverse the characters in a string
    */

    revStr(str) {
        return [...str].reverse().join('');
    }

    /*
     * Perform an exclusive-or of the corresponding bytes
     * in two byte arrays
     */
    xor(d, doff, s1, s1off, s2, s2off, len) {
        for (let i = 0; i < len; i++) {
            d[doff + i] = s1[s1off + i] ^ s2[s2off + i];
        }
    }

    /*
     * convert a big integer to a string under the radix @r with
     * length @m. If the string is longer than @m, the function fails.
     * if the string is shorter that @m, it is zero-padded to the left
    i: type bigInt
    r: int Radix
    m: length
    */

    str(m, r, i) {
        if (!this.custom_radix_str) {
            const s = i.toString(r);
            if (s.length > m) {
                throw new Error(errorMessages.StringExceeds);
            } else if (s.length < m) {
                return s.padStart(m, '0'); // TODO - This may not be safe if custom_radix_str[0] is not the '0' character
            }
            return s;
        }
        const s = Bn.bigint_get_str(this.custom_radix_str, i);
        if (s.length > m) {
            throw new Error(errorMessages.StringExceeds);
        } else if (s.length < m) {
            return s.padStart(m, this.custom_radix_str[0]);
        }
        return s;
    }

    /**
     * Encrypt a string, returning a cipher text using the same alphabet.
     *
     * The key, tweak parameters, and radix were all already set
     * by the initialization of the FF3_1 object.
     *
     * @param X   the plain text to be encrypted
     * @param twk the tweak used to perturb the encryption
     *
     * @return    the encryption of the plain text, the cipher text
     */
    encrypt(X, twk) {
        return this.cipher(X, twk, true);
    }

    /**
     * Decrypt a string, returning the plain text.
     *
     * The key, tweak parameters, and radix were all already set
     * by the initialization of the FF3_1 object.
     *
     * @param X   the cipher text to be decrypted
     * @param twk the tweak used to perturb the encryption
     *
     * @return    the decryption of the cipher text, the plain text
     */
    decrypter(X, twk) {
        return this.cipher(X, twk, false);
    }

    /**
     * Decrypt a string, returning the plain text.
     *
     * The key, tweak parameters, and radix were all already set
     * by the initialization of the FF3_1 object.
     *
     * @param X   the cipher text to be decrypted
     *
     * @return    the decryption of the cipher text, the plain text
     */
    decrypt(X) {
        return this.decrypter(X, null);
    }

    BigIntToByteArray(bn) {
        let hex = bn.toString(16);
        if (hex.length % 2) {
            hex = `0${hex}`;
        }
        const len = hex.length / 2;
        const u8 = new Uint8Array(len);

        let i = 0;
        let j = 0;
        while (i < len) {
            u8[i] = parseInt(`${hex[j]}${hex[j + 1]}`, 16);
            i += 1;
            j += 2;
        }
        return u8;
    }

    ByteArrayToBigInt(buf) {
        const hex = [];
        const u8 = this.Uint8Array.from(buf);

        u8.forEach((i) => {
            let h = i.toString(16);
            if (h.length % 2) { h = `0${h}`; }
            hex.push(h);
        });
        return bigInt(`0x${hex.join('')}`);
    }
}

module.exports = FFX;
